scopus-email-file-downloader/src/main/java/no/sikt/nva/scopus/email/ScopusEmailHandler.java

```diff
@@ -0,0 +1,139 @@
+
+    public static final int SINGLE_EXPECTED_RECORD = 0;
+    public static final String S3_URI_TEMPLATE = "s3://%s/%s";
+    private static final Logger logger = LoggerFactory.getLogger(ScopusEmailHandler.class);
+    private static final String ZIP_FILE_BUCKET = new Environment().readEnv("ZIP_FILE_BUCKET");
+    public static final String ERROR_STORING_SCOPUS_ZIP_FILE_FROM_URL = "Error storing file '%s' from url: '%s'";
+    public static final String SCOPUS_ZIP_FILE_STORED_FROM_URL = "File '%s' stored from url: '%s'";
+    public static final String ERROR_UNEXPECTED_STATUS_CODE = "Unexpected status code '%s' for request to uri";
+    public static final String REGEX_PATTERN_VALID_URL = "https://(sccontent-scudd-delivery-prod\\.s3\\.amazonaws\\.com"
+            + "/sccontent-scudd-delivery-prod/|localhost){1}[\\w.\\-/:#?=&;%~+]+";
+    private final S3Client s3Client;
+    private final HttpClient httpClient;
+
+    @JacocoGenerated
+    public ScopusEmailHandler() {
+        this(S3Driver.defaultS3Client().build(), HttpClient.newBuilder().build());
+    }
+
+    public ScopusEmailHandler(S3Client s3Client, HttpClient httpClient) {
+        this.s3Client = s3Client;
+        this.httpClient = httpClient;
+    }
+
+    @Override
+    public Void handleRequest(S3Event event, Context context) {
+        return attempt(() -> readFile(event))
+                .map(this::processEmailFile)
+                .orElseThrow(fail -> logErrorAndThrowException(fail.getException()));
+    }
+
+    private String readFile(S3Event event) {
+        var s3Driver = new S3Driver(s3Client, extractBucketName(event));
+        var fileUri = createS3BucketUri(event);
+        return s3Driver.getFile(UriWrapper.fromUri(fileUri).toS3bucketPath());
+    }
+
+    private String extractBucketName(S3Event event) {
+        return event.getRecords().get(SINGLE_EXPECTED_RECORD).getS3().getBucket().getName();
+    }
+
+    private String extractFilename(S3Event event) {
+        return event.getRecords().get(SINGLE_EXPECTED_RECORD).getS3().getObject().getKey();
+    }
+
+    private URI createS3BucketUri(S3Event s3Event) {
+        return URI.create(String.format(S3_URI_TEMPLATE, extractBucketName(s3Event), extractFilename(s3Event)));
+    }
+
+    private Void processEmailFile(String emailContents) {
+        Email email = EmailConverter.emlToEmailBuilder(emailContents).buildEmail();
+        extractValidUrlsFromEmail(email.getPlainText()).forEach(this::downloadFile);
+        return null;
+    }
+
+    private Set<URI> extractValidUrlsFromEmail(String emailContents) {
+        Set<URI> uris = new HashSet<>();
+        Pattern pattern = Pattern.compile(REGEX_PATTERN_VALID_URL, Pattern.CASE_INSENSITIVE);
+        Matcher matcher = pattern.matcher(emailContents);
+        while (matcher.find()) {
+            uris.add(URI.create(emailContents.substring(matcher.start(0),matcher.end(0))));
+        }
+        return uris;
+    }
+
+    private void downloadFile(URI uri) {
```

Comment:
There is an issue here, from what I understand. Given 3 URIs, where the second URI fails, the first URI is persisted, the second throws a fatal exception and the third is ignored. This gives us a state which means the input is partially processed, a state we probably cannot recover from (can we reliably resubmit the input and not create duplicates in s3? Does this have side effects if we do?). There are two options here:
- make the code transactional
- collect successes and failures and handle them in separate processes

Which you choose depends on what the requirements are.

Personally, I would also parallelise this.